File: working dir/widgets/inspector/Inspector.java
package inspector;

import java.io.File;
import java.nio.file.Path;
import java.util.Optional;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.css.PseudoClass;
import javafx.event.Event;
import javafx.scene.Node;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeView;
import sp.it.pl.gui.objects.tree.TreeItemsKt;
import sp.it.pl.layout.widget.Widget;
import sp.it.pl.layout.widget.controller.SimpleController;
import sp.it.pl.layout.widget.controller.io.Output;
import sp.it.pl.layout.widget.feature.FileExplorerFeature;
import sp.it.pl.layout.widget.feature.Opener;
import sp.it.pl.main.Widgets;
import sp.it.pl.util.graphics.drag.DragUtil;
import static javafx.css.PseudoClass.getPseudoClass;
import static javafx.scene.control.SelectionMode.MULTIPLE;
import static sp.it.pl.layout.widget.Widget.Group.APP;
import static sp.it.pl.util.graphics.Util.setAnchor;
import static sp.it.pl.util.graphics.UtilKt.expandAndSelect;
import static sp.it.pl.util.graphics.UtilKt.propagateESCAPE;

@Widget.Info(
    author = "Martin Polakovic",
    name = Widgets.INSPECTOR,
    description = "Inspects application as hierarchy. Displays windows, widgets,"
                + "file system and more. Allows editing if possible.",
    howto = "Available actions:\n"
          + "    Right click: Open context menu\n"
          + "    Double click file: Open file in native application\n"
          + "    Double click skin file: Apply skin on application\n"
          + "    Double click widget file: Open widget\n"
          + "    Drag & drop file: Explore file\n"
          + "    Drag & drop files: Explore files' first common parent directory\n",
    version = "0.8",
    year = "2015",
    group = APP
)
public class Inspector extends SimpleController implements FileExplorerFeature, Opener {

    private static final PseudoClass selectedPC = getPseudoClass("selected");
    private Node sel_node = null;
    private TreeView<Object> tree = TreeItemsKt.buildTreeView();
    private Output<Object> out_sel;

    public Inspector(Widget<?> widget) {
        super(widget);

        setAnchor(this, tree,0d);

        tree.getSelectionModel().setSelectionMode(MULTIPLE);
        tree.setCellFactory(TreeItemsKt::buildTreeCell);
        tree.setRoot(TreeItemsKt.treeApp());
        tree.getSelectionModel().selectedItemProperty().addListener((o,ov,nv) -> {
            Object oi = ov==null ? null : ov.getValue();
            Object ni = nv==null ? null : nv.getValue();

            out_sel.setValue(ni);

            // selected node highlighting
            if (sel_node!=null) {
                sel_node.pseudoClassStateChanged(selectedPC, false);
                sel_node.setStyle("");
                sel_node = null;
            }
            if (ni instanceof Node) {
                sel_node = (Node) ni;
                highlightNode(sel_node);
            }
        });
	    propagateESCAPE(tree);

        setOnDragOver(DragUtil.fileDragAcceptHandler);
        setOnDragDropped(e -> exploreFiles(DragUtil.getFiles(e)));

        // prevent scrolling event from propagating
        setOnScroll(Event::consume);
    }

    @Override
    public void init() {
        out_sel = outputs.create(widget.id,"Selected", Object.class, null);
        onClose.plusAssign(() -> unhighlightNode(sel_node));
    }

    @SuppressWarnings("unchecked")
    @Override
    public void exploreFile(File f) {
        TreeItem<File> root = (TreeItem) tree.getRoot().getChildren().get(3);

        // expand up to root + partitions
        tree.getRoot().setExpanded(true);
        root.setExpanded(true);

        // expand the most possible
        Path p = f.toPath().getRoot();
        Optional<TreeItem<File>> item = root.getChildren().stream().filter(i -> i.getValue().toString().contains(f.toPath().getRoot().toString())).findFirst();
        item.ifPresent(e -> e.setExpanded(true));
        ObjectProperty<TreeItem<File>> it = new SimpleObjectProperty<>(item.orElse(null));

        f.getAbsoluteFile().toPath().forEach(pth -> {
            if (it.get()!=null) {
                it.get().setExpanded(true);
                it.set(it.get().getChildren().stream().filter(i -> i.getValue().toString().contains(pth.toString())).findFirst().orElse(null));
            }
        });

        // expand if last==directory
        if (it.get()!=null && it.get().getValue().isDirectory()) it.get().setExpanded(true);

    }

    @Override
    public void open(Object data) {
        TreeItem<Object> item = TreeItemsKt.tree(data);
        tree.getRoot().getChildren().add(item);
        expandAndSelect(tree, item);
    }

    private static void highlightNode(Node n) {
        if (n==null) return;
        n.pseudoClassStateChanged(selectedPC, true);
        n.setStyle("-fx-background-color: rgba(90,200,200,0.2);");
    }

    private static void unhighlightNode(Node n) {
        if (n==null) return;
        n.pseudoClassStateChanged(selectedPC, false);
        n.setStyle("");
    }
}

File: working dir/widgets/inspector/Inspector.kt
package inspector

import javafx.css.PseudoClass.getPseudoClass
import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.control.SelectionMode.MULTIPLE
import javafx.scene.control.TreeItem
import javafx.scene.input.ScrollEvent
import javafx.util.Callback
import sp.it.pl.gui.objects.tree.buildTreeCell
import sp.it.pl.gui.objects.tree.buildTreeView
import sp.it.pl.gui.objects.tree.tree
import sp.it.pl.gui.objects.tree.treeApp
import sp.it.pl.layout.widget.Widget
import sp.it.pl.layout.widget.controller.SimpleController
import sp.it.pl.layout.widget.controller.io.Output
import sp.it.pl.layout.widget.feature.FileExplorerFeature
import sp.it.pl.layout.widget.feature.Opener
import sp.it.pl.main.APP
import sp.it.pl.main.Widgets
import sp.it.pl.main.rowHeight
import sp.it.pl.util.file.isAnyChildOf
import sp.it.pl.util.functional.ifIs
import sp.it.pl.util.graphics.drag.DragUtil
import sp.it.pl.util.graphics.expandToRootAndSelect
import sp.it.pl.util.graphics.layFullArea
import sp.it.pl.util.graphics.propagateESCAPE
import sp.it.pl.util.reactive.sync
import java.io.File

@Widget.Info(
        author = "Martin Polakovic",
        name = Widgets.INSPECTOR,
        description = "Inspects application as hierarchy. Displays windows, widgets,"+"file system and more. Allows editing if possible.",
        howto = ""
                +"Available actions:\n"
                +"    Right click: Open context menu\n"
                +"    Double click file: Open file in native application\n"
                +"    Double click skin file: Apply skin on application\n"
                +"    Double click widget file: Open widget\n"
                +"    Drag & drop file: Explore file\n"
                +"    Drag & drop files: Explore files' first common parent directory\n",
        version = "0.8",
        year = "2015",
        group = Widget.Group.APP
)
class Inspector(widget: Widget<*>): SimpleController(widget), FileExplorerFeature, Opener {
    private val tree = buildTreeView<Any>()
    private val outSelected: Output<Any?> = outputs.create(widget.id, "Selected", Any::class.java, null)

    init {
        this layFullArea tree.apply {
            selectionModel.selectionMode = MULTIPLE
            cellFactory = Callback { buildTreeCell(it) }
            onClose += APP.ui.font sync { fixedCellSize = it.rowHeight() }
            root = treeApp()
            selectionModel.selectedItemProperty().addListener { o, ov, nv ->
                val valueOld = ov?.value
                val valueNew = nv?.value
                if (valueOld is Node) valueOld.highlightNode(false)
                if (valueNew is Node) valueNew.highlightNode(true)
                outSelected.value = valueNew
            }
            propagateESCAPE()
        }

        onDragOver = DragUtil.fileDragAcceptHandler
        onDragDropped = EventHandler { exploreCommonFileOf(DragUtil.getFiles(it)) }
        onScroll = EventHandler<ScrollEvent> { it.consume() } // prevent scrolling event from propagating

        onClose += { outSelected.value.ifIs<Node> { it.highlightNode(false) } }
    }

    override fun exploreFile(f: File) {
        val fileRoot = tree.root.children[3]
        fun TreeItem<Any>.findChild() = children.find { it.value?.let { it is File && (f==it || f.isAnyChildOf(it)) }==true }
        generateSequence(fileRoot) { it.findChild() }.drop(1).lastOrNull()?.expandToRootAndSelect(tree)
    }

    override fun open(data: Any) {
        val item = tree(data)
        tree.root.children += item
        tree.expandToRootAndSelect(item)
    }

    override fun focus() = tree.requestFocus()

        companion object {

        private val selected = getPseudoClass("selected")

        private fun Node.highlightNode(value: Boolean) {
            pseudoClassStateChanged(selected, value)
            style = if (value) "-fx-background-color: rgba(90,200,200,0.2);" else ""
        }

    }
}